### PR TITLE
Add a TraceConfig for active TraceParams and allow changing the active TraceParams.

### PR DESCRIPTION
**Pull request by bogdandrutu**
* Moved the TraceParams as a nested class in the TraceConfig.

### CODE REVIEW

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+   * @return the active {@code TraceParams}.
+   */
+  public abstract TraceParams getActiveTraceParams();
+
+  /**
+   * Updates the active {@link TraceParams} permanently.
+   *
+   * @param traceParams the new active {@code TraceParams}.
+   */
+  public abstract void permanentlyUpdateActiveTraceParams(TraceParams traceParams);
```
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Permanently makes it sound like the TraceParams cannot be set again.  Would it be better to name the methods updateActiveTraceParams and temporarilyUpdateActiveTraceParams?

File: core/src/main/java/com/google/instrumentation/trace/TraceComponent.java [hunk 2]
```diff
@@ -56,6 +56,14 @@
    */
   public abstract TraceExporter getTraceExporter();
 
+  /**
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/TraceComponent.java]: The comment at line 20 should also mention this class (and Clock and TraceExporter)

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.auto.value.AutoValue;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Global configuration of the trace service. This allows users to change configs for the default
+ * sampler, maximum events to be kept, etc (see {@link TraceParams} for details).
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: nit: . after etc.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      /**
+       * Sets the global default max number of {@link Link} entries per {@link Span}.
+       *
+       * @param maxNumberOfLinks the global default max number of {@link Link} entries per {@link
+       *     Span}. It must be positive otherwise {@link #build()} will throw an exception.
+       * @return this.
+       */
+      public abstract Builder setMaxNumberOfLinks(int maxNumberOfLinks);
+
+      public abstract TraceParams autoBuild();
```
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: I don't think this should be public.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+  /**
+   * Returns the active {@code TraceParams}.
+   *
+   * @return the active {@code TraceParams}.
+   */
+  public abstract TraceParams getActiveTraceParams();
+
+  /**
+   * Updates the active {@link TraceParams} permanently.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Better: "Permanently update the active {@link TraceParams}."

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+   *
+   * @param traceParams the new active {@code TraceParams}.
+   */
+  public abstract void permanentlyUpdateActiveTraceParams(TraceParams traceParams);
+
+  /**
+   * Updates the active {@link TraceParams} temporary for {@code durationNs} nanoseconds.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: What is the use case for temporary updates?

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+  static TraceConfig getNoopTraceConfig() {
+    return noopTraceConfig;
+  }
+
+  /** Class that holds global trace parameters. */
+  @AutoValue
+  @Immutable
+  public abstract static class TraceParams {
+    // These values are the default values for all the global parameters.
+    // TODO(aveitch): Change this when a rate/probability sampler will be available.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: s/will be/is/

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        TraceParams.builder()
+            .setSampler(DEFAULT_SAMPLER)
+            .setMaxNumberOfAttributes(DEFAULT_SPAN_MAX_NUM_ATTRIBUTES)
+            .setMaxNumberOfAnnotations(DEFAULT_SPAN_MAX_NUM_ANNOTATIONS)
+            .setMaxNumberOfNetworkEvents(DEFAULT_SPAN_MAX_NUM_NETWORK_EVENTS)
+            .setMaxNumberOfLinks(DEFAULT_SPAN_MAX_NUM_LINKS)
+            .build();
+
+    /**
+     * Returns the global default {@code Sampler}. Used if no {@code Sampler} provided in {@link
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: s/provided/is provided/

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+       */
+      public abstract Builder setMaxNumberOfLinks(int maxNumberOfLinks);
+
+      public abstract TraceParams autoBuild();
+
+      /**
+       * Builds and returns a {@code TraceParams} with the desired values.
+       *
+       * @return a {@code TraceParams} with the desired values.
+       * @throws NullPointerException if the sampler is null.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: IllegalStateException does not seem like the correct exception "Signals that a method has been invoked at an illegal or inappropriate time." I would replace both of these with:
@throws IllegalArgumentException if sampler is {@code null} or any of the max numbers are not positive.

File: core/src/main/java/com/google/instrumentation/trace/TraceComponent.java (hunk 2 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceComponent.java]: done.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java (hunk 3 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Done.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java (hunk 5 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Done.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java (hunk 6 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: debugging and want to change the sampling probability for 1 min for example.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java (hunk 7 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Done.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java (hunk 8 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Done.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java (hunk 9 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Done.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java (hunk 1 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Done.

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java (hunk 4 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Done.

**Comment by bogdandrutu** [comment]
PTAL

File: core/src/main/java/com/google/instrumentation/trace/TraceConfig.java [hunk 10]
```diff
@@ -175,22 +175,23 @@ private static Builder builder() {
...
+      abstract TraceParams autoBuild();
 
       /**
        * Builds and returns a {@code TraceParams} with the desired values.
        *
        * @return a {@code TraceParams} with the desired values.
        * @throws NullPointerException if the sampler is null.
-       * @throws IllegalStateException if any of the max numbers are not positive.
+       * @throws IllegalArgumentException if sampler is {@code null} or any of the max numbers are
+       *     not positive.
```
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: The Javadoc lists two exception types for a null sampler now.

Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/TraceConfig.java]: Done. I think keeping NullPointerException for sampler makes more sense.

**Pull request merged**